Tasks/Common/RemoteDeployer/SessionHelper.ps1

```diff
@@ -0,0 +1,116 @@
+function ConnectTo-RemoteMachine {
+    [CmdletBinding()]
+    Param (
+        [string] $targetMachineName,
+        [pscredential] $credential,
+        [string] $protocol,
+        [string] $authentication,
+        [string] $sessionName,
+        [ValidateRange(2,10)]
+        [int] $maxRetryLimit = 3
+    )
+
+    Trace-VstsEnteringInvocation $MyInvocation
+    try {
+        $retryCount = 0;
+        $isConnectionComplete = $false
+        $useSsl = ($protocol -eq "https")
+        $machineDetails = Get-MachineDetails -MachineName $targetMachineName -useSsl:$useSsl
+        $newPsSessionCommand = Get-NewPSSessionCommand -machineDetails $machineDetails `
+                                                       -authentication $authentication `
+                                                       -sessionName $sessionName `
+                                                       -useSsl:$useSsl `
+                                                       -NoCredential:($credential -eq $null)
+
+        while ($retryCount -lt $maxRetryLimit) {
+            Write-Verbose "Trying to establish connection: Attempt #$($retryCount + 1)"
+            $session = (Invoke-Expression $newPsSessionCommand)
+            
+            foreach ($sessionError in $sessionErrors) {
+                Write-Verbose $("New-PSSession Error: " + $sessionError.Exception.Message)
+            }
+
+            if($session -ne $null) {
+                $isConnectionComplete = $true
+                break;
+            }
+
+            $retryCount++
+            Start-Sleep -Seconds 30
+        }
+
+        if ($isConnectionComplete) {
+            Write-Verbose "Connection established to computer:'$($machineDetails.ComputerName)' port:'$($machineDetails.WSManPort)'"
+            Write-Host $(Get-VstsLocString -Key "RemoteDeployer_ConnectedMachines" -ArgumentList $targetMachineName)
+            return $session    
+        } else {
+            foreach ($sessionError in $sessionErrors) {
+                Write-VstsTaskError -Message $sessionError.Exception.Message -ErrCode "PS_TM_UnableToCreatePSSession"
+            }
+            Write-VstsSetResult -Result 'Failed' -Message $(Get-VstsLocString -Key "RemoteDeployer_NotConnectedMachines" -ArgumentList $targetMachineName)
+        }
+    } finally {
+        Trace-VstsLeavingInvocation $MyInvocation
+    }
+}
+
+function Get-NewPSSessionCommand {
+    [CmdletBinding()]
+    Param (
+        [hashtable] $machineDetails,
+        [string] $authentication,
```

Comment:
should we have a default value for authentication here as I believe we always pass 'Default' and it is not a user input

Reply:
for psontargetmachines it is an input. But keeping 'default' here as the default value here makes sense. I will update it.